edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/TwinHandler.cs

```diff
@@ -0,0 +1,354 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Text.RegularExpressions;
+    using System.Threading.Channels;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Edge.Hub.Core;
+    using Microsoft.Azure.Devices.Edge.Hub.Core.Device;
+    using Microsoft.Azure.Devices.Edge.Hub.Core.Identity;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Extensions.Logging;
+
+    public class TwinHandler : ITwinHandler, IMessageConsumer, IMessageProducer
+    {
+        const string TwinGetDevice = "$edgehub/+/twin/get/#";
+        const string TwinGetModule = "$edgehub/+/+/twin/get/#";
+        const string TwinUpdateDevice = "$edgehub/+/twin/reported/#";
+        const string TwinUpdateModule = "$edgehub/+/+/twin/reported/#";
+
+        const string TwinGetPublishPattern = @"^\$edgehub/(?<id1>[^/\+\#]+)(/(?<id2>[^/\+\#]+))?/twin/get/\?\$rid=(?<rid>.+)";
+        const string TwinUpdatePublishPattern = @"^\$edgehub/(?<id1>[^/\+\#]+)(/(?<id2>[^/\+\#]+))?/twin/reported/\?\$rid=(?<rid>.+)";
+
+        const string TwinSubscriptionForResultsPattern = @"^\$edgehub/(?<id1>[^/\+\#]+)(/(?<id2>[^/\+\#]+))?/twin/res/\#$";
+        const string TwinSubscriptionForPatchPattern = @"^\$edgehub/(?<id1>[^/\+\#]+)(/(?<id2>[^/\+\#]+))?/twin/desired/\#$";
+
+        const string TwinResultDevice = "$edgehub/{0}/twin/res/{1}/?$rid={2}";
+        const string TwinResultModule = "$edgehub/{0}/{1}/twin/res/{2}/?$rid={3}";
+
+        const string DesiredUpdateDevice = "$edgehub/{0}/twin/desired/?$version={1}";
+        const string DesiredUpdateModule = "$edgehub/{0}/{1}/twin/desired/?$version={2}";
+
+        static readonly string[] subscriptions = new[] { TwinGetDevice, TwinGetModule, TwinUpdateDevice, TwinUpdateModule };
+
+        static readonly SubscriptionPattern[] subscriptionPatterns = new SubscriptionPattern[]
+                                                                         {
+                                                                            new SubscriptionPattern(TwinSubscriptionForResultsPattern, DeviceSubscription.TwinResponse),
+                                                                            new SubscriptionPattern(TwinSubscriptionForPatchPattern, DeviceSubscription.DesiredPropertyUpdates)
+                                                                         };
+        readonly Channel<ProcessingInfo> notifications;
```

Comment:
It is interesting that we need the Channel pattern for the twins (which is not expected to be high throughput fast messaging communication). How slow is the SendGetTwin and UpdateReportedProperties?